pilot/adapter/serviceregistry/aggregate/mesh_view.go

```diff
@@ -0,0 +1,406 @@
+		serviceInstanceLabels: make(nameValueKeysMap),
+	}
+}
+
+// A resource key representing a mesh service resource
+// Format for service: [service name][platformRegistry]
+// There can only be one service object per platformRegistry
+// TODO: this will change with multi-cluster to be one service object per platformRegistry per cluster
+func BuildServiceKey(r *Registry, s *model.Service) resourceKey {
+	return resourceKey("[" + s.Hostname + "][" + string(r.Name) + "]")
+}
+
+// Format for service instance: [service name][hex value of IP address][hex value of port number]
+// Within the mesh there can be exactly one endpoint for a service with the combination of
+// IP address and port
+func BuildServiceInstanceKey(i *model.ServiceInstance) resourceKey {
+	return resourceKey("[" + i.Service.Hostname + "][" + getIPHex(i.Endpoint.Address) + "][" + getPortHex(i.Endpoint.Port) + "]")
+}
+
+func getIPHex(address string) string {
+	ip := net.ParseIP(address)
+	return hex.EncodeToString(ip)
+}
+
+func getPortHex(port int) string {
+	pb := []byte{byte((port >> 8) & 0xFF), byte(port & 0xFF)}
+	return hex.EncodeToString(pb)
+}
+
+// A few labels are treated differently to ensure
+// compatibility between various numeric and IP text
+// values that are otherwise identical, ex: 08880 and 8080
+// or 10.1.1.3 and ::ffff:10.1.1.3
+func resourceLabelFromNameValue(label string, value *string) resourceLabel {
+	switch label {
+	case labelServiceVIP:
+		fallthrough
+	case labelInstanceIP:
+		ipHex := getIPHex(*value)
+		return resourceLabel{label, &ipHex}
+	case labelInstancePort:
+		var port int
+		fmt.Sscanf(*value, "%d", &port)
+		portHex := getPortHex(port)
+		return resourceLabel{label, &portHex}
+	}
+	return resourceLabel{label, value}
+}
+
+func resourceLabelsFromModelLabels(lc model.Labels) resourceLabels {
+	rl := make(resourceLabels, len(lc))
+	i := 0
+	for k, v := range lc {
+		rl[i] = resourceLabelFromNameValue(k, &v)
+		i++
+	}
+	return rl
+}
+
+func resourceLabelsFromNameValues(label string, values []string) resourceLabels {
+	rl := make(resourceLabels, len(values))
+	for idx, v := range values {
+		rl[idx] = resourceLabelFromNameValue(label, &v)
+	}
+	return rl
+}
+
+func (r *Registry) HandleService(s *model.Service, e model.Event) {
+	k := BuildServiceKey(r, s)
+	r.MeshView.handleService(k, s, e)
+}
+
+func (r *Registry) HandleServiceInstance(i *model.ServiceInstance, e model.Event) {
+	k := BuildServiceInstanceKey(i)
+	r.MeshView.handleServiceInstance(k, i, e)
+}
+
+// AddRegistry adds registries into the aggregated MeshResourceView
+func (v *MeshResourceView) AddRegistry(registry Registry) {
+	// Create bidirectional associations between the MeshView and
+	// the registry being added
+	registry.MeshView = v
+	v.registries = append(v.registries, registry)
+}
+
+// Services lists services from all platforms
+func (v *MeshResourceView) Services() ([]*model.Service, error) {
+	lbls := resourceLabelsForName(labelServiceName)
```

Comment:
`labels` would probably suffice and be more readable

Reply:
@GregHanson I'm assuming "resourceLabels" vs "resourceLabelsForName" would be OK with you. Rationale for "resource" is disambiguate between labels supplied by xDS which need to support empty values and resourceLabels which is internally used by xDS and can have nil values to enable querying for everything matching a the label name (irrespective of value)

Reply:
I think he was referring to the local variable `lbls`? The function name looks good to me as is.